rcore/flash.c

```diff
@@ -158,54 +180,55 @@ static void _fs_init()
     uint8_t saw_page_in_outer_space = 0;
     for (pg = 0; pg < REGION_FS_N_PAGES; pg++)
     {
-        _fs_read_page_hdr(pg, &hdr);
-        if (hdr.v_0x5001 == 0xFFFF) {
+        _fs_read_page_hdr(pg, &buffer);
+        if (hdr->v_0x5001 == 0xFFFF) {
             if (!saw_blank_page)
                 KERN_LOG("flash", APP_LOG_LEVEL_ERROR, "this filesystem has a blank page in it... hmm...");
             saw_blank_page = 1;
             continue;
         }
-        if (hdr.v_0x5001 != 0x5001) {
-            KERN_LOG("flash", APP_LOG_LEVEL_ERROR, "page %d has bad header version 0x%04x; I give up", pg, hdr.v_0x5001);
+        if (hdr->v_0x5001 != 0x5001) {
+            KERN_LOG("flash", APP_LOG_LEVEL_ERROR, "page %d has bad header version 0x%04x; I give up", pg, hdr->v_0x5001);
             _fs_valid = 0;
             return;
         }
-        if (hdr.status == 0xFE && lastpg != -1) {
+        if (hdr->status == 0xFE && lastpg != -1) {
             lastpg = pg;
         }
-        if ((lastpg != -1) && FLASHFLAG(hdr.empty, HDR_EMPTY_ALLOCATED)) {
+        if ((lastpg != -1) && FLASHFLAG(hdr->empty, HDR_EMPTY_ALLOCATED)) {
             if (!saw_page_in_outer_space)
                 KERN_LOG("flash", APP_LOG_LEVEL_ERROR, "page %d is marked as allocated, but page %d had the last page marker... hmm...", pg, lastpg);
             saw_page_in_outer_space = 1;
         }
         
         /* The rest of the checks only apply to an allocated page. */
-        if (!FLASHFLAG(hdr.empty, HDR_EMPTY_ALLOCATED))
+        if (!FLASHFLAG(hdr->empty, HDR_EMPTY_ALLOCATED))
             continue;
-        if (!FLASHFLAG(hdr.status, HDR_STATUS_FILE_START))
+
+        if (FLASHFLAG(hdr->status, HDR_STATUS_FILE_CONT))
+            page_flags[pg].state = PageStateFileCont;
+
+        if (!FLASHFLAG(hdr->status, HDR_STATUS_FILE_START))
             continue;
-        if (!FLASHFLAG(hdr.status, HDR_STATUS_DEAD) && hdr.st_create_complete) {
+
+        if (!FLASHFLAG(hdr->status, HDR_STATUS_DEAD) && hdr->st_create_complete) {
             KERN_LOG("flash", APP_LOG_LEVEL_ERROR, "page %d creation not complete; I can't deal with this; go boot PebbleOS to clean up first", pg);
             _fs_valid = 0;
             return;
         }
-        if (FLASHFLAG(hdr.status, HDR_STATUS_DEAD) && hdr.st_delete_complete) {
+        if (FLASHFLAG(hdr->status, HDR_STATUS_DEAD) && hdr->st_delete_complete) {
             KERN_LOG("flash", APP_LOG_LEVEL_ERROR, "page %d deletion not complete; I can't deal with this; go boot PebbleOS to clean up first", pg);
             _fs_valid = 0;
             return;
         }
```

Comment:
do an integrity check if MAX_FILENAME_LEN doesn't come from something in PebbleOS?